File: week1/js/data/books.js
let bookArray = [];
let bookID = 0;

// Added a couple of initial books
bookArray.push({
    "id": bookID++,
    "title": "lets make it count", 
    "author": "pointy pencil", 
    "year": 1992, 
    "genre": "sci-fi"
});
bookArray.push({
    "id": bookID++,
    "title": "lizard people", 
    "author": "tinfoil hat", 
    "year": 2017, 
    "genre": "sci-fi"
});
bookArray.push({
    "id": bookID++,
    "title": "Divergent Aperture Harmonics", 
    "author": "Domadagr", 
    "year": 2025, 
    "genre": "biography"
});

const getBooks = ((req) => {
    const { page = 1, limit = 10 } = req.query;
    const startIndex = (page - 1) * limit;
    const endIndex = startIndex + parseInt(limit);

    return {
        totalBooks: bookArray.length,
        books: bookArray.slice(startIndex, endIndex)
    };
});

const addBook = ((book) => {
    const newBook = { "id": bookID++, ...book };
    bookArray.push(newBook);
    return newBook;
});

const verifyPayload = ((req, res, next) => {
    const { title, author, year, genre } = req.body;
    if (!title || !author || !year || !genre) {
        return res.status(400).send({ error: "All fields required: Title, Author, Year, Genre" });
    }
    next();
});

module.exports = { getBooks, addBook, verifyPayload };

File: week1/py/data/books.py
from pydantic import BaseModel
from typing import Optional

bookID = 3


class Book(BaseModel):
    id: Optional[int] = None,
    title: str
    author: str
    year: int


books = [
    {"id": 1, "title": "The Great Adventure", "author": "John Doe", "year": 2005},
    {"id": 2, "title": "Learning FastAPI", "author": "Jane Smith", "year": 2021},
]


def getbooks():
    return books


def add_book(book):
    global bookID
    new_book = {"id": bookID, **book.dict(exclude={"id"})}  # Exclude 'id' from the request
    books.append(new_book)
    bookID += 1
    return new_book


def update_book(book_id, updated_book):
    for i, book in enumerate(books):
        if book['id'] == book_id:
            books[i] = updated_book
            return updated_book
    return None


def get_book_by_id(book_id):
    for i, book in enumerate(books):
        if book['id'] == book_id:
            return book
    return None


def remove_book_by_id(book_id):
    for i, book in enumerate(books):
        if book['id'] == book_id:
            books.pop(i)
            return book_id
    return None

